### PR TITLE
HHH-12858 + HHH-13432 integration overrides during JPA bootstrap ought to overr

### PR DESCRIPTION
**Pull request by dreab8**
ide all logically related settings.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/internal/util/NullnessHelper.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	 * @return The first non-empty value, or null if all values were empty
+	 */
+	public static <T> T coalesce(T... values) {
+		if ( values == null ) {
+			return null;
+		}
+		for ( T value : values ) {
+			if ( value != null ) {
+				if ( String.class.isInstance( value ) ) {
+					if ( StringHelper.isNotEmpty( (String) value ) ) {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/internal/util/NullnessHelper.java]: Minor nitpicking, as I've recently being removing the use of "unnecessary helpers":

when you know already that "value" is not null, there is no reason to use the `StringHelper`. Just go with `((String) value).isEmpty()` ?

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/util/NullnessHelper.java]: Thanks @Sanne , I'll change it.

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/internal/util/NullnessHelper.java]: I guess you missed this one then - I literally copied this as is :D

But, +1

**Comment by dreab8** [comment]
squashed and applied upstream

File: hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java [hunk 2]
```diff
@@ -567,22 +543,446 @@ else if ( keyString.startsWith( COLLECTION_CACHE_PREFIX ) ) {
...
+		final String inverseEmfKey;
+		if ( isJta ) {
+			emfKey = JPA_JTA_DATASOURCE;
+			inverseEmfKey = JPA_NON_JTA_DATASOURCE;
+		}
+		else {
+			emfKey = JPA_NON_JTA_DATASOURCE;
+			inverseEmfKey = JPA_JTA_DATASOURCE;
+		}
+		mergedSettings.configurationValues.put( emfKey, dataSourceRef );
```
Comment (jwgmeligmeyling) [inline hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java]: It seems that due to this change, settings from the `persistence.xml` cannot be cleared anymore, only overridden:

i.e.:

```java
properties.put(AvailableSettings.JPA_JTA_DATASOURCE, null);
entityManagerFactory = Persistence.createEntityManagerFactory("athena", properties);
```

Used to work, but now throws:

```
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyDataSource(EntityManagerFactoryBuilderImpl.java:895)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.normalizeDataAccess(EntityManagerFactoryBuilderImpl.java:758)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.normalizeSettings(EntityManagerFactoryBuilderImpl.java:561)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.mergeSettings(EntityManagerFactoryBuilderImpl.java:485)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:205)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:175)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:76)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilder(HibernatePersistenceProvider.java:171)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:119)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:61)
	at org.hibernate.jpa.HibernatePersistenceProvider.createEntityManagerFactory(HibernatePersistenceProvider.java:50)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:79)
	at com.pallasathenagroup.entitymanagers.AbstractEntityManagerTest.initializeEntityManagerFactory(AbstractEntityManagerTest.java:83)
```

Comment (jwgmeligmeyling) [reply hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java]: Ah nvm , it seems that setting the properties to `null` is simply not required anymore, as the settings are already cleared when setting `AvailableSettings.JPA_JDBC_URL`

**Pull request closed without merging**